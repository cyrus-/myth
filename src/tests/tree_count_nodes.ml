let s = "
type nat =
  | O
  | S of nat

type tree =
  | Leaf
  | Node of tree * nat * tree

let rec sum (n1:nat) (n2:nat) : nat =
  match n1 with
  | O -> n2
  | S (n1) -> S (sum n1 n2)
;;

let tree_count_nodes : tree -> nat |>
  { Leaf => 0
  | Node(Leaf, 0, Leaf) => 1
  | Node(Node(Leaf, 0, Leaf), 0, Leaf) => 2
  | Node(Leaf, 0, Node(Leaf, 0, Leaf)) => 2
  | Node(Node(Leaf, 0, Node(Leaf, 0, Leaf)), 0, Leaf) => 3
  | Node(Leaf, 0, Node(Leaf, 0, Node(Leaf, 0, Leaf))) => 3
  } = ?"
