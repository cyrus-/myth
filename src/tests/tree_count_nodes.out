(* Generated Thu Nov 13 00:35:54 2014 *)
let tree_count_nodes : tree -> nat =
  let rec f1 (t1:tree) : nat =
    match t1 with
      | Leaf -> O
      | Node (t2, n1, t3) -> S (sum (f1 t2) (f1 t3))
  in
    f1
;;
